docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,380 @@
+   # boilerplate to generate tree type and instantiate a tree ...
+
+   @ti.kernel
+   def foo(tr: tree_t):
+     x = ti.static(tr.get_field(x_handle))  # 1
+     for i in x:
+       x[i] = i * 2.0
+   ```
+
+   Note that this design requires that part of the kernel (1) being evaluated inside Python. It also pulls in the global variable `x_handle`, which kind of violates our initial goal. We could require that `x_handle` is passed into the kernel as an argument. But maybe it's fine just to view that as a trivial Python constant?
+
+## AoS vs SoA
+
+Two composite types require the switch between AoS vs SoA, `ti.Matrix` and `ti.Struct`.
+
+AoS is quite straightforward. One can just use the composite type as the `dtype` of the field. For example:
+
+```py
+builder = ti.SNodeTreeBuilder()
+
+builder.add_field(dtype=ti.vec3, name='x')  # ti.vec3 is a vector of 3 ti.f32's
+builder.dense(ti.i, 8).place('x')
+tree_t = x.build()
+```
+
+For SoA, things get a bit trickier. The **current approach** is to treat each compopnent of the composite type as a standalone scalar Taichi field. In the example below, we have to manually place the underlying 3 components of `x` separately.
+
+```py
+# Current way (as of v1.0.1) of doing SoA in Taichi
+x = ti.Vector.field(3, ti.f32)
+for f in x._get_field_members():  # `x` consists three scalar f32 fields
+  ti.root.dense(ti.ij).place(f)
+```
+
+This introduces confusion at several spots:
+
+1. Type is not purely decided by `dtype`, but also by how the field is placed.
+2. It introduces the notion of "nested field", which Taichi doesn't currently have a good abstraction for. Because of this, it is quite complicated to apply certain kind of optimizations for a composite-typed field. For example, vectorized load/save consumes the same bandwidth as scalar ops on certain platforms. Without a good abstraction, the checking for whether a matrix field is AoS or SoA has to be spread across different passes in CHI IR.
+
+If we further think about the problem, SoA `x` is not really a field. Instead, it is a *grouped view* of three individual scalar fields. This view provides matrix operations, which won't make sense for each individual scalar field.
+
+In addition, because type is currently coupled with Taichi field definition, a Taichi field has to be implemented as individual fields in order to support the SoA scenario. Once we switch to the type builder pattern, we can control how the type is constructed first, and choose the field implementation later.
+
+If we want to make it explicit that this is a *field view*, we can do the following:
+
+```py
+builder = ti.SNodeTreeBuilder()
+builder.add_field(dtype=ti.f32, name='v0')
+builder.add_field(dtype=ti.f32, name='v1')
+builder.add_field(dtype=ti.f32, name='v2')
+for v in ['v0', 'v1', 'v2']:
+  builder.tree().dense(ti.ij, 4).place(v)
+
+# Checks that
+# 1. `components` and `dtype` are compatible.
+# 2. All the fields in `components` are homogeneous in their SNode hierarchy.
```

Comment:
I think checking that all components are having the same shape is enough.

Reply:
Hmmm if we say vec3 implies homogeneous elements, it's fair to say all the element fields must have identical snode hierarchy? I kinda feel enforcing this in the new api until there's a valid request might save us some time worrying about the edge cases.

Reply:
"homogeneous" itself doesn't have a clear meaning, and the criteria can vary significantly. However, the additional check and additional restrictions on user code don't seem to benefit us, especially when we decide to avoid manipulating backend pointers in CHI IR.

Reply:
When I wrote homogeneous, I was thinking of https://github.com/taichi-dev/taichi/issues/3810

Reply:
I think this is a fairly small point. Not enforcing this actually reduces our development efforts. In the long run though, not enforcing this could still result in certain "unexpected" use cases. I'd like to refer to Hyrum's Law here...

> With a sufficient number of users of an API, it does not matter what you promise in the contract: all observable behaviors of your system will be depended on by somebody.

Reply:
If we enforce it we might be able to hear from users who has a valid use case and share with us. But if not we'll always have to assume someone relies on this indeterministic use case :P

Reply:
Yep #3810 is exactly checking "homogeneous" with a relatively relaxed criteria. And behavior like #3810 indeed looks a bit weird to me as a user: why does an API work only when we use a bunch of them in certain patterns?

Reply:
Are we talking about vectors or structs here? They deserve different handling IMHO. I don't think homogeneous is necessary for struct views. For matrices, however, it's equally weird why someone would want vectors to take on different sparsity.

Reply:
> If we enforce it we might be able to hear from users who has a valid use case and share with us. But if not we'll always have to assume someone relies on this indeterministic use case :P

This sounds like a good reason for enforcing it!

> They deserve different handling IMHO.

I agree.

Reply:
Fixed